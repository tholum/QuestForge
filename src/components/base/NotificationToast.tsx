import * as React from "react"
import { cva, type VariantProps } from "class-variance-authority"
import { 
  CheckCircle2, 
  AlertCircle, 
  Info, 
  X, 
  AlertTriangle,
  Zap,
  Trophy,
  Target
} from "lucide-react"

import { cn } from "@/lib/utils"
import { Button } from "./Button"

const toastVariants = cva(
  "group pointer-events-auto relative flex w-full items-center justify-between space-x-4 overflow-hidden rounded-md border p-4 pr-8 shadow-lg transition-all data-[swipe=cancel]:translate-x-0 data-[swipe=end]:translate-x-[var(--radix-toast-swipe-end-x)] data-[swipe=move]:translate-x-[var(--radix-toast-swipe-move-x)] data-[swipe=move]:transition-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[swipe=end]:animate-out data-[state=closed]:fade-out-80 data-[state=closed]:slide-out-to-right-full data-[state=open]:slide-in-from-top-full data-[state=open]:sm:slide-in-from-bottom-full",
  {
    variants: {
      variant: {
        default: "border bg-background text-foreground",
        success: "border-green-200 bg-green-50 text-green-800 dark:border-green-800 dark:bg-green-950 dark:text-green-300",
        error: "border-red-200 bg-red-50 text-red-800 dark:border-red-800 dark:bg-red-950 dark:text-red-300",
        warning: "border-yellow-200 bg-yellow-50 text-yellow-800 dark:border-yellow-800 dark:bg-yellow-950 dark:text-yellow-300",
        info: "border-blue-200 bg-blue-50 text-blue-800 dark:border-blue-800 dark:bg-blue-950 dark:text-blue-300",
        achievement: "border-purple-200 bg-gradient-to-r from-purple-50 to-pink-50 text-purple-800 dark:border-purple-800 dark:from-purple-950 dark:to-pink-950 dark:text-purple-300",
      },
      size: {
        sm: "text-sm p-3",
        default: "text-sm p-4",
        lg: "text-base p-5",
      },
      position: {
        "top-left": "",
        "top-right": "",
        "top-center": "",
        "bottom-left": "",
        "bottom-right": "",
        "bottom-center": "",
      },
    },
    defaultVariants: {
      variant: "default",
      size: "default",
      position: "top-right",
    },
  }
)

const toastIcons = {
  default: Info,
  success: CheckCircle2,
  error: AlertCircle,
  warning: AlertTriangle,
  info: Info,
  achievement: Trophy,
} as const

export interface INotificationToastProps extends
  React.HTMLAttributes<HTMLDivElement>,
  VariantProps<typeof toastVariants> {
  title: string
  description?: string
  icon?: React.ReactNode
  action?: {
    label: string
    onClick: () => void
  }
  onClose?: () => void
  closeable?: boolean
  duration?: number
  progress?: number
  showProgress?: boolean
  persistent?: boolean
  soundEnabled?: boolean
  vibrate?: boolean
}

const NotificationToast = React.forwardRef<HTMLDivElement, INotificationToastProps>(
  ({
    className,
    variant = "default",
    size,
    position,
    title,
    description,
    icon: customIcon,
    action,
    onClose,
    closeable = true,
    duration = 5000,
    progress,
    showProgress = false,
    persistent = false,
    soundEnabled = false,
    vibrate = false,
    ...props
  }, ref) => {
    const [isVisible, setIsVisible] = React.useState(true)
    const [progressValue, setProgressValue] = React.useState(100)
    const timerRef = React.useRef<NodeJS.Timeout>()
    const progressTimerRef = React.useRef<NodeJS.Timeout>()

    const IconComponent = toastIcons[variant as keyof typeof toastIcons]
    const icon = customIcon || <IconComponent className="h-5 w-5 flex-shrink-0" />

    // Auto dismiss functionality
    React.useEffect(() => {
      if (persistent || duration <= 0) return

      // Progress bar animation
      if (showProgress) {
        const progressInterval = 50
        const progressStep = (progressInterval / duration) * 100
        
        const updateProgress = () => {
          setProgressValue(prev => {
            const newValue = prev - progressStep
            if (newValue <= 0) {
              handleClose()
              return 0
            }
            return newValue
          })
        }

        progressTimerRef.current = setInterval(updateProgress, progressInterval)
      }

      // Auto close timer
      timerRef.current = setTimeout(() => {
        if (!persistent) {
          handleClose()
        }
      }, duration)

      return () => {
        if (timerRef.current) clearTimeout(timerRef.current)
        if (progressTimerRef.current) clearInterval(progressTimerRef.current)
      }
    }, [duration, persistent, showProgress])

    // Handle sounds and vibration
    React.useEffect(() => {
      if (soundEnabled) {
        // Play notification sound (implement based on your audio system)
        // playNotificationSound(variant)
      }

      if (vibrate && "vibrate" in navigator) {
        const pattern = variant === "error" ? [100, 50, 100] : [50]
        navigator.vibrate(pattern)
      }
    }, [soundEnabled, vibrate, variant])

    const handleClose = () => {
      setIsVisible(false)
      onClose?.()
      
      // Clean up timers
      if (timerRef.current) clearTimeout(timerRef.current)
      if (progressTimerRef.current) clearInterval(progressTimerRef.current)
    }

    const handleMouseEnter = () => {
      // Pause auto-dismiss on hover
      if (timerRef.current) clearTimeout(timerRef.current)
      if (progressTimerRef.current) clearInterval(progressTimerRef.current)
    }

    const handleMouseLeave = () => {
      // Resume auto-dismiss when not hovering
      if (!persistent && duration > 0) {
        const remainingTime = (progressValue / 100) * duration
        
        if (showProgress) {
          const progressInterval = 50
          const progressStep = (progressInterval / remainingTime) * progressValue
          
          progressTimerRef.current = setInterval(() => {
            setProgressValue(prev => {
              const newValue = prev - progressStep
              if (newValue <= 0) {
                handleClose()
                return 0
              }
              return newValue
            })
          }, progressInterval)
        }

        timerRef.current = setTimeout(handleClose, remainingTime)
      }
    }

    if (!isVisible) return null

    return (
      <div
        ref={ref}
        className={cn(toastVariants({ variant, size, position }), className)}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        role="alert"
        aria-live={variant === "error" ? "assertive" : "polite"}
        {...props}
      >
        {/* Progress bar */}
        {showProgress && (
          <div className="absolute top-0 left-0 right-0 h-1 bg-black/10 dark:bg-white/10">
            <div
              className="h-full bg-current transition-all duration-75 ease-linear"
              style={{ width: `${progressValue}%` }}
            />
          </div>
        )}

        {/* Content */}
        <div className="flex items-start space-x-3 flex-1">
          {icon && (
            <div className="flex-shrink-0 pt-0.5">
              {icon}
            </div>
          )}
          
          <div className="flex-1 space-y-1">
            <div className="font-semibold leading-none tracking-tight">
              {title}
            </div>
            {description && (
              <div className="text-sm opacity-90 leading-relaxed">
                {description}
              </div>
            )}
          </div>
        </div>

        {/* Action button */}
        {action && (
          <div className="flex-shrink-0">
            <Button
              variant="ghost"
              size="sm"
              onClick={action.onClick}
              className="h-auto p-2 text-current hover:bg-current/10"
            >
              {action.label}
            </Button>
          </div>
        )}

        {/* Close button */}
        {closeable && (
          <Button
            variant="ghost"
            size="icon-sm"
            onClick={handleClose}
            className="absolute top-2 right-2 h-6 w-6 text-current hover:bg-current/10 focus:ring-current/20"
          >
            <X className="h-3 w-3" />
            <span className="sr-only">Close</span>
          </Button>
        )}
      </div>
    )
  }
)

NotificationToast.displayName = "NotificationToast"

// Toast container component
export const ToastContainer: React.FC<{
  position?: VariantProps<typeof toastVariants>["position"]
  maxToasts?: number
  children: React.ReactNode
}> = ({ 
  position = "top-right", 
  maxToasts = 5,
  children 
}) => {
  const positionClasses = {
    "top-left": "fixed top-4 left-4 z-50",
    "top-right": "fixed top-4 right-4 z-50",
    "top-center": "fixed top-4 left-1/2 transform -translate-x-1/2 z-50",
    "bottom-left": "fixed bottom-4 left-4 z-50",
    "bottom-right": "fixed bottom-4 right-4 z-50",
    "bottom-center": "fixed bottom-4 left-1/2 transform -translate-x-1/2 z-50",
  }

  return (
    <div className={cn(positionClasses[position], "space-y-2 max-w-sm w-full")}>
      {children}
    </div>
  )
}

// Toast manager hook
export const useToast = () => {
  const [toasts, setToasts] = React.useState<Array<{
    id: string
    props: INotificationToastProps
  }>>([])

  const addToast = (props: Omit<INotificationToastProps, "onClose">) => {
    const id = Math.random().toString(36).substr(2, 9)
    
    setToasts(prev => [...prev, {
      id,
      props: {
        ...props,
        onClose: () => removeToast(id)
      }
    }])

    return id
  }

  const removeToast = (id: string) => {
    setToasts(prev => prev.filter(toast => toast.id !== id))
  }

  const clearAllToasts = () => {
    setToasts([])
  }

  // Convenience methods
  const success = (title: string, description?: string) => 
    addToast({ variant: "success", title, description })

  const error = (title: string, description?: string) => 
    addToast({ variant: "error", title, description })

  const warning = (title: string, description?: string) => 
    addToast({ variant: "warning", title, description })

  const info = (title: string, description?: string) => 
    addToast({ variant: "info", title, description })

  const achievement = (title: string, description?: string) => 
    addToast({ variant: "achievement", title, description, duration: 8000 })

  return {
    toasts,
    addToast,
    removeToast,
    clearAllToasts,
    success,
    error,
    warning,
    info,
    achievement,
  }
}

export { NotificationToast, toastVariants }