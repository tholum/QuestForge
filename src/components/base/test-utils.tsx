import { render, RenderOptions } from '@testing-library/react'
import { ReactElement } from 'react'
import { vi } from 'vitest'

// Mock IntersectionObserver
global.IntersectionObserver = vi.fn(() => ({
  disconnect: vi.fn(),
  observe: vi.fn(),
  unobserve: vi.fn(),
}))

// Mock ResizeObserver
global.ResizeObserver = vi.fn(() => ({
  disconnect: vi.fn(),
  observe: vi.fn(),
  unobserve: vi.fn(),
}))

// Mock matchMedia
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // deprecated
    removeListener: vi.fn(), // deprecated
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
})

// Mock navigator.vibrate for QuickAddButton tests
Object.defineProperty(navigator, 'vibrate', {
  writable: true,
  value: vi.fn(),
})

// Mock requestAnimationFrame
global.requestAnimationFrame = vi.fn(cb => setTimeout(cb, 16))
global.cancelAnimationFrame = vi.fn()

// Provider wrapper for components that might need context
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <div className="bg-background text-foreground">
      {children}
    </div>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }

// Common test utilities
export const mockGoal = {
  id: 'test-goal-1',
  title: 'Test Goal',
  description: 'A test goal for components',
  isCompleted: false,
  isPaused: false,
  isCancelled: false,
  dueDate: new Date(Date.now() + 86400000), // Tomorrow
  startDate: new Date(),
  priority: 'medium' as const,
  category: 'fitness' as const,
  progress: 45,
  createdAt: new Date(),
  updatedAt: new Date(),
}

export const mockUser = {
  id: 'test-user-1',
  name: 'Test User',
  email: 'test@example.com',
  level: 5,
  xp: 1250,
  nextLevelXp: 1500,
  streak: 12,
  achievements: 8,
}

export const createMockGoal = (overrides?: Partial<typeof mockGoal>) => ({
  ...mockGoal,
  ...overrides,
})

// Accessibility test helpers
export const axeMatchers = {
  toHaveNoViolations: expect.toHaveNoViolations,
}

// Animation helpers for components with animations
export const waitForAnimation = (duration = 500) => 
  new Promise(resolve => setTimeout(resolve, duration))

// Component testing utilities
export const getByTextContent = (container: HTMLElement, text: string) => {
  return Array.from(container.querySelectorAll('*')).find(
    element => element.textContent === text
  )
}

export const queryByTextContent = (container: HTMLElement, text: string) => {
  return Array.from(container.querySelectorAll('*')).find(
    element => element.textContent === text
  ) || null
}

// Toast testing utilities (since toasts might be portaled)
export const findToastByText = (text: string) => {
  // Look in document body since toasts might be portaled
  const elements = Array.from(document.body.querySelectorAll('*'))
  return elements.find(el => el.textContent?.includes(text))
}

// Form testing utilities
export const fillAndSubmitForm = async (
  user: ReturnType<typeof import('@testing-library/user-event').default.setup>,
  fields: Record<string, string>
) => {
  for (const [label, value] of Object.entries(fields)) {
    const field = screen.getByLabelText(new RegExp(label, 'i'))
    await user.clear(field)
    await user.type(field, value)
  }
}

export * from '@testing-library/user-event'