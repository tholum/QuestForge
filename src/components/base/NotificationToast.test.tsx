import { render, screen, waitFor, act } from './test-utils'
import userEvent from '@testing-library/user-event'
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { NotificationToast, useToast, ToastContainer } from './NotificationToast'
import { Trophy } from 'lucide-react'

// Mock timers for auto-dismiss testing
vi.useFakeTimers()

describe('NotificationToast', () => {
  beforeEach(() => {
    vi.clearAllTimers()
  })

  it('renders toast with title and description', () => {
    render(
      <NotificationToast
        title="Test Toast"
        description="This is a test notification"
      />
    )

    expect(screen.getByText('Test Toast')).toBeInTheDocument()
    expect(screen.getByText('This is a test notification')).toBeInTheDocument()
  })

  it('applies variant styles correctly', () => {
    const { rerender } = render(
      <NotificationToast
        variant="success"
        title="Success"
        description="Operation completed"
      />
    )

    const toast = screen.getByText('Success').closest('[role="alert"]')
    expect(toast).toHaveClass('border-green-200', 'bg-green-50')

    rerender(
      <NotificationToast
        variant="error"
        title="Error"
        description="Something went wrong"
      />
    )

    expect(toast).toHaveClass('border-red-200', 'bg-red-50')
  })

  it('renders with custom icon', () => {
    render(
      <NotificationToast
        title="Achievement"
        description="You earned a badge!"
        icon={<Trophy data-testid="trophy-icon" />}
      />
    )

    expect(screen.getByTestId('trophy-icon')).toBeInTheDocument()
  })

  it('renders action button and handles click', async () => {
    const mockAction = vi.fn()
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime })

    render(
      <NotificationToast
        title="Toast with Action"
        description="Click the action button"
        action={{
          label: "Take Action",
          onClick: mockAction
        }}
      />
    )

    const actionButton = screen.getByRole('button', { name: /take action/i })
    expect(actionButton).toBeInTheDocument()

    await user.click(actionButton)
    expect(mockAction).toHaveBeenCalledTimes(1)
  })

  it('renders close button and handles close', async () => {
    const mockClose = vi.fn()
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime })

    render(
      <NotificationToast
        title="Closeable Toast"
        description="This toast can be closed"
        closeable
        onClose={mockClose}
      />
    )

    const closeButton = screen.getByRole('button', { name: /close/i })
    expect(closeButton).toBeInTheDocument()

    await user.click(closeButton)
    expect(mockClose).toHaveBeenCalledTimes(1)
  })

  it('hides close button when closeable is false', () => {
    render(
      <NotificationToast
        title="Non-closeable Toast"
        description="This toast cannot be closed"
        closeable={false}
      />
    )

    expect(screen.queryByRole('button', { name: /close/i })).not.toBeInTheDocument()
  })

  it('shows progress bar when showProgress is true', () => {
    render(
      <NotificationToast
        title="Progress Toast"
        description="This toast shows progress"
        showProgress
        duration={5000}
      />
    )

    // Look for progress bar element
    const progressBar = document.querySelector('.h-1')
    expect(progressBar).toBeInTheDocument()
  })

  it('auto-dismisses after specified duration', async () => {
    const mockClose = vi.fn()

    render(
      <NotificationToast
        title="Auto-dismiss Toast"
        description="This will auto-dismiss"
        duration={3000}
        onClose={mockClose}
      />
    )

    expect(mockClose).not.toHaveBeenCalled()

    // Fast-forward time
    act(() => {
      vi.advanceTimersByTime(3000)
    })

    await waitFor(() => {
      expect(mockClose).toHaveBeenCalledTimes(1)
    })
  })

  it('does not auto-dismiss when persistent is true', async () => {
    const mockClose = vi.fn()

    render(
      <NotificationToast
        title="Persistent Toast"
        description="This will not auto-dismiss"
        persistent
        duration={1000}
        onClose={mockClose}
      />
    )

    act(() => {
      vi.advanceTimersByTime(2000)
    })

    expect(mockClose).not.toHaveBeenCalled()
  })

  it('pauses auto-dismiss on hover', async () => {
    const mockClose = vi.fn()
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime })

    render(
      <NotificationToast
        title="Hover Toast"
        description="Hover pauses auto-dismiss"
        duration={2000}
        onClose={mockClose}
      />
    )

    const toast = screen.getByText('Hover Toast').closest('[role="alert"]')!

    // Start timer
    act(() => {
      vi.advanceTimersByTime(1000)
    })

    // Hover to pause
    await user.hover(toast)

    // Advance time - should not dismiss yet
    act(() => {
      vi.advanceTimersByTime(2000)
    })

    expect(mockClose).not.toHaveBeenCalled()

    // Un-hover to resume
    await user.unhover(toast)

    // Now it should dismiss
    act(() => {
      vi.advanceTimersByTime(2000)
    })

    await waitFor(() => {
      expect(mockClose).toHaveBeenCalledTimes(1)
    })
  })

  it('applies correct size classes', () => {
    const { rerender } = render(
      <NotificationToast
        title="Small Toast"
        size="sm"
      />
    )

    const toast = screen.getByText('Small Toast').closest('[role="alert"]')
    expect(toast).toHaveClass('text-sm', 'p-3')

    rerender(
      <NotificationToast
        title="Large Toast"
        size="lg"
      />
    )

    expect(toast).toHaveClass('text-base', 'p-5')
  })

  it('has proper accessibility attributes', () => {
    render(
      <NotificationToast
        variant="error"
        title="Error Message"
        description="This is an error"
      />
    )

    const toast = screen.getByRole('alert')
    expect(toast).toHaveAttribute('role', 'alert')
    expect(toast).toHaveAttribute('aria-live', 'assertive')
  })

  it('has polite aria-live for non-error variants', () => {
    render(
      <NotificationToast
        variant="info"
        title="Info Message"
        description="This is info"
      />
    )

    const toast = screen.getByRole('alert')
    expect(toast).toHaveAttribute('aria-live', 'polite')
  })

  // Test useToast hook
  describe('useToast hook', () => {
    const TestComponent = () => {
      const toast = useToast()

      return (
        <div>
          <button onClick={() => toast.success('Success!', 'It worked!')}>
            Add Success
          </button>
          <button onClick={() => toast.error('Error!', 'Something failed')}>
            Add Error
          </button>
          <button onClick={() => toast.clearAllToasts()}>
            Clear All
          </button>
          
          <ToastContainer position="top-right">
            {toast.toasts.map(({ id, props }) => (
              <NotificationToast key={id} {...props} />
            ))}
          </ToastContainer>
        </div>
      )
    }

    it('adds and manages toasts', async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime })

      render(<TestComponent />)

      // Add success toast
      await user.click(screen.getByText('Add Success'))
      expect(screen.getByText('Success!')).toBeInTheDocument()
      expect(screen.getByText('It worked!')).toBeInTheDocument()

      // Add error toast
      await user.click(screen.getByText('Add Error'))
      expect(screen.getByText('Error!')).toBeInTheDocument()
      expect(screen.getByText('Something failed')).toBeInTheDocument()

      // Should have both toasts
      expect(screen.getAllByRole('alert')).toHaveLength(2)
    })

    it('clears all toasts', async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime })

      render(<TestComponent />)

      // Add toasts
      await user.click(screen.getByText('Add Success'))
      await user.click(screen.getByText('Add Error'))

      expect(screen.getAllByRole('alert')).toHaveLength(2)

      // Clear all
      await user.click(screen.getByText('Clear All'))

      expect(screen.queryByRole('alert')).not.toBeInTheDocument()
    })

    it('removes individual toasts', async () => {
      const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime })

      render(<TestComponent />)

      await user.click(screen.getByText('Add Success'))
      await user.click(screen.getByText('Add Error'))

      expect(screen.getAllByRole('alert')).toHaveLength(2)

      // Close one toast
      const closeButtons = screen.getAllByRole('button', { name: /close/i })
      await user.click(closeButtons[0])

      await waitFor(() => {
        expect(screen.getAllByRole('alert')).toHaveLength(1)
      })
    })
  })

  // ToastContainer tests
  describe('ToastContainer', () => {
    it('applies position classes correctly', () => {
      const { rerender } = render(
        <ToastContainer position="top-left">
          <div data-testid="toast-content">Toast</div>
        </ToastContainer>
      )

      const container = screen.getByTestId('toast-content').parentElement
      expect(container).toHaveClass('fixed', 'top-4', 'left-4')

      rerender(
        <ToastContainer position="bottom-center">
          <div data-testid="toast-content">Toast</div>
        </ToastContainer>
      )

      expect(container).toHaveClass('bottom-4', 'left-1/2')
    })

    it('limits number of toasts when maxToasts is specified', () => {
      render(
        <ToastContainer maxToasts={2}>
          <div data-testid="toast-1">Toast 1</div>
          <div data-testid="toast-2">Toast 2</div>
          <div data-testid="toast-3">Toast 3</div>
        </ToastContainer>
      )

      // Should render all children (limiting would be handled by the hook logic)
      expect(screen.getByTestId('toast-1')).toBeInTheDocument()
      expect(screen.getByTestId('toast-2')).toBeInTheDocument()
      expect(screen.getByTestId('toast-3')).toBeInTheDocument()
    })
  })

  // Snapshot test
  it('matches snapshot', () => {
    const { container } = render(
      <NotificationToast
        variant="achievement"
        title="Achievement Unlocked!"
        description="You've completed your first goal!"
        icon={<Trophy />}
        action={{
          label: "View Badge",
          onClick: () => {}
        }}
        showProgress
        duration={5000}
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})