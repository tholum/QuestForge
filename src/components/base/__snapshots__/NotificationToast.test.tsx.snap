// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationToast > matches snapshot 1`] = `
<div
  class="bg-background text-foreground"
>
  <div
    aria-live="polite"
    class="group pointer-events-auto relative flex w-full items-center justify-between space-x-4 overflow-hidden rounded-md border shadow-lg transition-all data-[swipe=cancel]:translate-x-0 data-[swipe=end]:translate-x-[var(--radix-toast-swipe-end-x)] data-[swipe=move]:translate-x-[var(--radix-toast-swipe-move-x)] data-[swipe=move]:transition-none data-[state=open]:animate-in data-[state=closed]:animate-out data-[swipe=end]:animate-out data-[state=closed]:fade-out-80 data-[state=closed]:slide-out-to-right-full data-[state=open]:slide-in-from-top-full data-[state=open]:sm:slide-in-from-bottom-full border-purple-200 bg-gradient-to-r from-purple-50 to-pink-50 text-purple-800 dark:border-purple-800 dark:from-purple-950 dark:to-pink-950 dark:text-purple-300 text-sm p-4"
    role="alert"
  >
    <div
      class="absolute top-0 left-0 right-0 h-1 bg-black/10 dark:bg-white/10"
    >
      <div
        class="h-full bg-current transition-all duration-75 ease-linear"
        style="width: 100%;"
      />
    </div>
    <div
      class="flex items-start space-x-3 flex-1"
    >
      <div
        class="flex-shrink-0 pt-0.5"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-trophy"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 14.66v1.626a2 2 0 0 1-.976 1.696A5 5 0 0 0 7 21.978"
          />
          <path
            d="M14 14.66v1.626a2 2 0 0 0 .976 1.696A5 5 0 0 1 17 21.978"
          />
          <path
            d="M18 9h1.5a1 1 0 0 0 0-5H18"
          />
          <path
            d="M4 22h16"
          />
          <path
            d="M6 9a6 6 0 0 0 12 0V3a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1z"
          />
          <path
            d="M6 9H4.5a1 1 0 0 1 0-5H6"
          />
        </svg>
      </div>
      <div
        class="flex-1 space-y-1"
      >
        <div
          class="font-semibold leading-none tracking-tight"
        >
          Achievement Unlocked!
        </div>
        <div
          class="text-sm opacity-90 leading-relaxed"
        >
          You've completed your first goal!
        </div>
      </div>
    </div>
    <div
      class="flex-shrink-0"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap font-medium transition-all duration-200 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive focus:scale-[0.98] active:scale-[0.96] select-none hover:text-accent-foreground dark:hover:bg-accent/50 text-sm gap-1.5 has-[>svg]:px-2.5 rounded-md h-auto p-2 text-current hover:bg-current/10"
        data-loading="false"
        data-slot="button"
      >
        <span
          class="flex-1"
        >
          View Badge
        </span>
      </button>
    </div>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all duration-200 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive focus:scale-[0.98] active:scale-[0.96] select-none hover:text-accent-foreground dark:hover:bg-accent/50 size-8 rounded-md absolute top-2 right-2 h-6 w-6 text-current hover:bg-current/10 focus:ring-current/20"
      data-loading="false"
      data-slot="button"
    >
      <span
        class="flex-1"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-x h-3 w-3"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 6 6 18"
          />
          <path
            d="m6 6 12 12"
          />
        </svg>
        <span
          class="sr-only"
        >
          Close
        </span>
      </span>
    </button>
  </div>
</div>
`;
